### ISSUE TITLE
[Feature]: A strategy for managing user configuration

### ISSUE THREAD

**Issue by rebuilt** [ChristianChiarulli/LunarVim#758]
### Feature motivation

Users are confused about the best way to manage configurations.  The current build of LunarVim allows for user configuration from one file.  However, some users might prefer to split up their configurations into multiple files.  

Additionally, users have asked for more plugins included with LunarVim.  

@ChristianChiarulli  You've brought up the idea of creating a marketplace for plugins

### Describe the expected solution

I have created a possible solution here: 
https://github.com/rebuilt/lv-user-configs

The lv-user-configs project is a way to manage LunarVim configuration files. The approach is to use a git submodule to separate out the two concerns: user-specific configuration from LunarVim core configuration. This means a user can create their own configuration files in their own repos. Changes will be tracked in their own git submodule. 

Having a separate repo also enables users to share configurations with one another. It means we might also create a marketplace for different modes: web development, script-writing, document creation, etc.

Modes can easily be switched by changing a symlink to a different folder. As long as it contains a init.lua file, it will work

**Comment by rebuilt** [comment]
Hm.  Maybe I should change the name of the repository from lv-user-configs to lunar-modules?

**Comment by rebuilt** [comment]
If anyone wants to share their own personal configs for others to use, please create a pull request on this repo.

**Comment by brymer-meneses** [comment]
Would it be better if we just create a separate folder like ~/.config/lunarvim for example to store all the user configurations. And just add logic inside core lunarvim to load the files inside ~/.config/lunarvim ?

**Comment by brymer-meneses** [comment]
That way we won’t have to use git submodules which makes the process simpler in my opinion.

**Comment by rebuilt** [comment]
You can keep your repo in an exterior folder as well.  You just have to change the symlinks.  And now that you mention it, that should probably be the default.  We don't have to add logic to core LunarVim.

```
git clone https://github.com/rebuilt/lv-user-config.git ~/.config/lunarvim
ln -s ~/.config/lunarvim/current/init.lua lv-config.lua
```

**Comment by rebuilt** [comment]
Sorry, let me correct the symbolic link commands

**Comment by brymer-meneses** [comment]
We could also automate the creation of the symlink and the boilerplate config file using the installer.

**Comment by rebuilt** [comment]
Ok,  This is how you would do it.

```
git clone https://github.com/rebuilt/lv-user-config.git ~/.config/lunarvim
ln -s ~/.config/nvim/lua/lv-user-config/current/init.lua ~/.config/nvim/lv-config.lua
ln -s ~/.config/lunarvim/ ~/.config/nvim/lua/lv-user-config

```

**Comment by rebuilt** [comment]
And then all your lunarvim configuration lives in ~/.config/lunarvim

**Comment by brymer-meneses** [comment]
I was thinking of adding logic inside core neovim to make this process streamlined but I guess the creation of the symlinks could be done using the installer.

**Comment by brymer-meneses** [comment]
If we make it so that `~/.config/lunarvim` will be the default directory for user configurations, we will just need to add the following line to core lunarvim.

```
vim.cmd [[ luafile ~/.config/lunarvim ]] 
```

Which IMO simplifies this whole process.

**Comment by brymer-meneses** [comment]
We won’t have to create symlinks and we could just point `~/.config/lunarvim/init.lua` to other user configuration directories which could reside on the `~/.config/lunarvim ` directory

**Comment by rebuilt** [comment]
That would work well for a single configuration.  But it's really powerful to be able to swap to a different configuration just by changing a symlink.  That way you can easily try out a different plugin suite.  I imagine one day adding example folders for each major language.  Or for different purposes like thesis writing

**Comment by brymer-meneses** [comment]
We could do this as well with this setup,

For example, suppose we have the following directories in `~/.config/lunarvim`
```
thesis-writing/
coding/
init.lua
```

We could just add the following line to `~/config/lunarvim/init.lua` to switch to `thesis-writing` mode.
```
vim.cmd [[ luafile ~/.config/lunarvim/thesis-writing/init.lua ]]
```

**Comment by rebuilt** [comment]
Would it be possible to write a telescope-based picker that could switch modes for you?  When you select an entry from the list of options, would you be able to modify which files get sourced?  I can imagine doing it by changing a symlink, but if you have to modify the source of the configuration files, I think it might be harder.   I really don't know the answer to this.  Can you see a way to do it?

**Comment by rebuilt** [comment]
Or maybe that kind of picker would make more sense as a command line utility?

**Comment by brymer-meneses** [comment]
We could use https://github.com/famiu/nvim-reload to dynamically source the configs, but it wouldn’t really change the default at runtime. For that I think we would need this https://www.tutorialspoint.com/lua/lua_file_io.html . I’m new to lua so idk.

**Comment by brymer-meneses** [comment]
> Or maybe that kind of picker would make more sense as a command line utility?

It would really be cool to use telescope though.

**Comment by rebuilt** [comment]
I added a command line script to change modes.  

To use, run:

```
./mode.sh
```

**Comment by gnmearacaun** [comment]
Thanks for setting this up. Not only does it make it easier to share configs while we're converting vim settings to lua. 

We could hop to a working config when the current one is broken due to syntax errors or unmerged conflicts.

I would name it config/ and then it will appear at the top of the folder (using ranger) and the first to autocomplete on bash/zsh commandline. 
Otherwise, it could be in utils/ where lv-user-config would be easier to get to.

On topic, it would be nice if lua/default-config.lua, lua/keymappings.lua, lua/plugins.lua, lua/settings.lua were all in lua/defaut-config/ for the same reason.

**Comment by Tastyep** [comment]
First, thank you rebuilt for this issue and making things move forward in the right direction !
Regarding how the plugin directory should be linked to the project, I think that @brymer-meneses's solution is slightly better as it would allow having a configuration variable to specify the path.

Also, Since you evoked the marketplace idea, wouldn't it be even better to provide a command (:PluginInstall for example) which would fetch the plugin and update the config directly ?

**Comment by rebuilt** [comment]
@Tastyep Thanks for the great ideas.  It would be really cool to have a PluginInstall command.  And yes, adding support for a custom path seems like a basic thing that should be supported.

I'm on a trip until Tuesday of next week.  When I get back, I'll see what I can do about supporting a configurable path through an environment variable.  Then I'll work on a technique to install plugins via some sort of :PluginInstall command.

This doesn't end the conversation about how configuration files should be handled but my repo is a stopgap measure that works today without modifying the internal functioning of LunarVim.  Not adding any more features to the internals of LunarVim will help us release 1.0 faster.

@gnmearacaun Those are some good ideas about having fallback profiles and rethinking the directory structure of configuration files.  For the moment though, I'll be prioritizing the things I mentioned above.

**Comment by rebuilt** [comment]
This strategy no longer works since lv-config is being pcalled

**Comment by gnmearacaun** [comment]
What does it mean for managing profiles in a folder?

**Comment by rebuilt** [comment]
@ChristianChiarulli Is there a plan for managing profiles in a folder?  I'm not sure it can be done through lv-config.lua since it is getting pcalled.  But maybe I'm wrong.  

Can we put another entry point into init.lua for this purpose?  We can check for the existence of lv-user/init.lua and load it if it's there.  It would have to be loaded with the 'luafile' command for this to work.  This would allow for requiring any number of files from the lv-user folder.

**Comment by ChristianChiarulli** [comment]
@rebuilt I would really like to have a whole directory rather than just a file.

**Comment by rebuilt** [comment]
This is no longer needed since the user configs were moved to `.config/lvim`.  That directory can now be its own git repository.

**Comment by aniketgm** [comment]
I just started with Neovim few weeks back and the first things I did was to have separate configs. This is how I have it with different files. Again I'm not a lua expert (new to lua and learning) but managed somehow to keep things separate. https://github.com/aniketgm/dotfiles/tree/main/.config/nvim

@rebuilt, appreciate the effort for having configs at diff locations, which I want as well. @ChristianChiarulli, can we have a user being given such flexibility as my example config above.

So, one thing I tried is that, I tried to extend config as follows `~/.config/lvim/myconfig/keymaps.lua` and added the following line inside it:
```
lvim.keys.normal_mode["<F3>"] = "<cmd>NvimTreeFindFileToggle<cr>"
```
and
inside `~/.config/lvim/config.lua:` I added:
```
require('myconfig.keymaps')
```
But this is giving me following error:
```
19:32:55 [WARN ] lvim: "Invalid configuration: /home/aniketgm/.config/lvim/config.lua:22: module 'myconfig.keymaps' not found:\n\tno field package.preload['
myconfig.keymaps']
```
